### ISSUE TITLE
Better handling for unwrapping reference to Result (or Option)

### ISSUE THREAD

**Issue by xliiv** [rust-lang/rust#48121]
Having such simplified code:

```rust
fn main() {
    let result: Result<String, String> = Ok("ok".to_string());
    let ref_result = &result;
    let x = ref_result.unwrap();
}
```
gives:

```
error[E0507]: cannot move out of borrowed content
 --> src/main.rs:4:13
  |
4 |     let x = ref_result.unwrap();
  |             ^^^^^^^^^^ cannot move out of borrowed content

error: aborting due to previous error

error: Could not compile `playground`.
```

Thanks to irc Kimundi & Amarnath I got the solution & expanation:

```rust
// use:
let x = ref_result.as_ref().unwrap();
```

### Explanation:
* `unwrap` requires full ownership because it moves values
    * see `self` in [signature](https://doc.rust-lang.org/core/result/enum.Result.html#method.unwrap)?
* `ref_result` is only borrowing, (`ref_result` is a reference, a reference means borrowing)

Using `.as_ref()` works because:
It creates a brand new `Result` with references to inner value, like:
from initial:

    `&Result<T, V>`

to brand new one:

    `Result<&T, &V>`

See implementation of `Result.as_ref` [here](https://doc.rust-lang.org/src/core/result.rs.html#400)

If i got this correctly, using `unwrap` on `&Result` will always be corrupted.
My question is, could this be handled internally somehow by Rust?
And/Or better compilation error would be awesome for noobs like me :)
I'm not sure what the message should look like but the current one is a bit general.

**Comment by ExpHP** [comment]
> If i got this correctly, using unwrap on `&Result` will always be corrupted.

You are correct, though there is one exception for the case where both `T` and `E` impl `Copy`.  In this case, `Result<T, E>` impls `Copy`, and thus calling a `self` function on a `&Result<T, E>` becomes allowed.

> My question is, could this be handled internally somehow by Rust?

This is no trivial matter, and is something that would probably require an RFC and require lots of discussion (and probably additions to the language) to make happen.  Consider `Option`, which is in a similar boat:

```rust
impl<T> Option<T> {
    pub fn as_ref(&self) -> Option<&T> {
        match *self {
            Option::Some(ref x) => Option::Some(x),
            Option::None => Option::None,
        }
    }
    
    pub fn unwrap(self) -> T {
        match self {
            Option::Some(x) => x,
            Option::None => panic!("unwrap on None"),
        }
    }
    
    pub fn unwrap_or_default(self) -> T
    where T: Default, {
        match self {
            Option::Some(x) => x,
            Option::None => Default::default(),
        }
    }
}
```

Both `unwrap()` and `unwrap_or_default()` are functions with the signature `fn(Option<T>) -> T`, but only one of them can used together with with `as_ref()` to behave like a `fn(&Option<T>) -> &T`.  Hence there would need to be something else in the signature to reflect this difference in capability; some way to indicate that `unwrap()` can be made polymorphic over ownership versus references.

(well, okay; technically the `where T: Default` bound is a difference between the signatures, and is in fact the key thing that makes `.as_ref().unwrap_as_default()` fail to compile; but attempting to apply this information in any sort of general fashion seems nontrivial, and I fear it could pose problems for specialization)

---

One final thought: It is unfortunate that default match binding modes will further blur the line between `&Result<T, E>` and `Result<&T, &E>`, potentially making issues like this feel even more subtle to beginners. But this was already anticipated...

**Comment by xliiv** [comment]
>  but only one of them can used together with with as_ref() to behave like a fn(&Option<T>) -> &T. Hence there would need to be something else in the signature to reflect this difference in capability; some way to indicate that unwrap() can be made polymorphic over ownership versus references.
> 
> (well, okay; technically the where T: Default bound is a difference between the signatures, and is in fact the key thing that makes .as_ref().unwrap_as_default() fail to compile; but attempting to apply this information in any sort of general fashion seems nontrivial, and I fear it could pose problems for specialization)

I don't get it. How the code for that would look like?

Anyway, If it's really non trivial case then a better error message will be good enough I think.
What do you @ExpHP think?

**Comment by ExpHP** [comment]
> Anyway, If it's really non trivial case then a better error message will be good enough I think.
> What do you @ExpHP think?

Seems possible.  The question is, what is the scope of the lint?  Just `Option` and `Result`? Does the compiler try to "guess" for the presence of methods like `as_ref()` based on their signature?  Should it also suggest `clone`, which is also frequently often the right solution for Option?

I will cc @estebank who has been prolific with diagnostic hints and may have better ideas.

**Comment by xliiv** [comment]
Now I got it, thanks for explanation.
Anyway kill me, but I'm not sure I like it yet. :) It's a bit parallel (a few meanings of `&?`) and implicit? I don't know.. .
However I  would happily see better compilation error, maybe I'm able even to contribute it?

**Comment by xliiv** [comment]
Maybe it could be also added to https://doc.rust-lang.org/error-index.html ?

**Comment by asquared31415** [comment]
Currently in 1.62.0 stable the output is now
```
error[[E0507]](https://doc.rust-lang.org/stable/error-index.html#E0507): cannot move out of `*ref_result` which is behind a shared reference
 --> src/main.rs:4:13
  |
4 |     let x = ref_result.unwrap();
  |             ^^^^^^^^^^^^^^^^^^^ move occurs because `*ref_result` has type `Result<String, String>`, which does not implement the `Copy` trait
  |
help: consider borrowing the `Result`'s content
  |
4 |     let x = ref_result.unwrap().as_ref();
  |                                +++++++++
  ```
  
However that suggestion is still wrong, and applying it will cause an error with regards to type inference which will lead you down a path that won't work.  The ideal suggestion is to put the `.as_ref()` before the `.unwrap()`.

**Comment by estebank** [comment]
Triage: [current output in beta is correct](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=9003e29b10878b303fcabcf90bc864f5):

```
error[E0507]: cannot move out of `*ref_result` which is behind a shared reference
 --> src/main.rs:4:13
  |
4 |     let x = ref_result.unwrap();
  |             ^^^^^^^^^^^--------
  |             |          |
  |             |          `*ref_result` moved due to this method call
  |             move occurs because `*ref_result` has type `Result<String, String>`, which does not implement the `Copy` trait
  |
note: this function takes ownership of the receiver `self`, which moves `*ref_result`
help: consider calling `.as_ref()` to borrow the type's contents
  |
4 |     let x = ref_result.as_ref().unwrap();
  |                        +++++++++
```

**Comment by Dylan-DPC** [comment]
Current error:

```
error[[E0507]](https://doc.rust-lang.org/stable/error_codes/E0507.html): cannot move out of `*ref_result` which is behind a shared reference
 --> src/main.rs:4:13
  |
4 |     let x = ref_result.unwrap();
  |             ^^^^^^^^^^^--------
  |             |          |
  |             |          `*ref_result` moved due to this method call
  |             help: consider calling `.as_ref()` or `.as_mut()` to borrow the type's contents
  |             move occurs because `*ref_result` has type `Result<String, String>`, which does not implement the `Copy` trait
  |
note: `Result::<T, E>::unwrap` takes ownership of the receiver `self`, which moves `*ref_result`
 --> /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/result.rs:1083:19
help: you can `clone` the value and consume it, but this might not be your desired behavior
  |
4 |     let x = ref_result.clone().unwrap();
  |                        ++++++++

For more information about this error, try `rustc --explain E0507`.
```
